### PR TITLE
API: Added NodeIteratorUtil, ImageRenderer, PublishingUtil, Utils, ImageScaler

### PR DESCRIPTION
**Pull request by kling90**
Added a couple of more API:s I've used in the last couple of weeks.

Not sure how to handle Utils since it can be used to get almost anything, but that would mean we have to write interface definitions for everything obtainable through Utils, that would lead to alot of duplication of documentation.
The primary way of should be to import the API:s and not getting them through Utils?

I've only added getImageScaler to Utils for now since it's the only way to get an instance of that API, I think?

### CODE REVIEW

**Comment by albinohrn** [comment]
I think you're correct in saying that the primary way should be to import the the API:s. However some of the API:s are stateful and importing them could lead to weird/unexpected behaviour.  That is because these external imports will be hoisted and only imported once. And if you're using them in multiple places of your app, you will have a shared instance of this stateful API. This is why we in our linter warns when importing stateful API:s.

You can get around this by using utils though, there you will get a new instance each time you're calling `utils.getImageRenderer()` for instance.

To summarize `getImageScaler` should be acquired through utils and so should `ImageRenderer` (and all other API:s ending in **renderer**).

**Comment by albinohrn** [comment]
Just to be clear. I don't mean that you have to add all the other renderers now! :)

**Comment by kling90** [comment]
Ah of course 😄 
When something is marked as "Requireable" in the docs I've alwasy assumed that's the way to import it. 
But it makes sense to go through Utils when talking about renderers,

I'll add ImageRenderer to Utils as well since I've added docs for that one already.

File: packages/api/src/server/Utils/index.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export default {
+  getImageScaler() {},
```
Comment (albinohrn) [inline packages/api/src/server/Utils/index.js:2]: Missing `getImageRenderer() {},`

File: packages/api/src/server/ImageRenderer/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+
+  /**
+  * Updates the renderer with a new image and a new description.
+  * @param {Node} anImageNode - the image (a node with primary node type sv:image)
+  * @param {string} aDescription - an image description, or null to clear any previous description 
+  */
+  update(anImageNode: Node, aDescription: string): void;
+}
+
+declare namespace imageRenderer {
```
Comment (albinohrn) [inline packages/api/src/server/ImageRenderer/index.d.ts:234]: In order to help prevent others from importing this API I think we should remove the `namespace`.

**Review by albinohrn — CHANGES_REQUESTED** [review]

File: packages/api/src/server/ImageRenderer/index.d.ts (hunk 2 above)
Comment (kling90) [reply packages/api/src/server/ImageRenderer/index.d.ts]: Agreed, should "non importable interfaces" be placed in another structure?
When removing the index.js the import is unresolvable (as it should be), but since the ImageRenderer is in the server folder it's still listed when writing the import, same with ImageScaler.

image

image

Comment (albinohrn) [reply packages/api/src/server/ImageRenderer/index.d.ts]: Yeah, that might be a good idea. But then the favorite issue of any developer arises - naming! :D 

I would go with a folder names stateful where we could put all of these APIs.

Comment (kling90) [reply packages/api/src/server/ImageRenderer/index.d.ts]: > There are only two hard things in Computer Science: cache invalidation and naming things.

Where to put the `stateful` folder? Directly under `./src` or in `./src/server`?
Either way it'll still be a suggestion when writing the import path, but builtins are also appearing even though they're not importable. Maybe I'm just overthinking 😄

Comment (albinohrn) [reply packages/api/src/server/ImageRenderer/index.d.ts]: Maybe we're both overthinking this. 😅 I'm back to keeping them in their "correct" place(`./src/server/ImageRenderer`). Then at least it is possible to import them although you'll get a lint warning. If you'd try to import from `stateful` it would break.

But I think we should document them as an interface that you acquire through `utils.get...`.

Comment (kling90) [reply packages/api/src/server/ImageRenderer/index.d.ts]: Should that be documented here? Or on developer.sitevision.se? 

It already says in the java docs that you should get it from Utils, 
https://developer.sitevision.se/webdav/files/apidocs/senselogic/sitevision/api/render/ImageRenderer.html

But I'm guilty of seeing `@Requireable(value="ImageRenderer")` and thinking "Oh I'll import it directly then!" 😅

Comment (albinohrn) [reply packages/api/src/server/ImageRenderer/index.d.ts]: We'll add more info to the lint warning I think. It'll still work in most cases, but will act weird in some. :)

**Review by albinohrn — APPROVED** [review]
